experimental/devices/mfrc522/mfrc522.go

```diff
@@ -0,0 +1,732 @@
+	},
+	waitLoop: [][]byte{
+		{commands.FIFODataReg, 0x26},
+		{commands.CommandReg, 0x0C},
+		{commands.BitFramingReg, 0x87},
+	},
+}
+
+func (r *Dev) writeCommandSequence(commands [][]byte) error {
+	for _, cmdData := range commands {
+		if err := r.devWrite(int(cmdData[0]), cmdData[1]); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Init initializes the RFID chip.
+func (r *Dev) Init() error {
+	if err := r.Reset(); err != nil {
+		return err
+	}
+	if err := r.writeCommandSequence(sequenceCommands.init); err != nil {
+		return err
+	}
+	return r.SetAntenna(true)
+}
+
+func (r *Dev) devWrite(address int, data byte) error {
+	newData := []byte{(byte(address) << 1) & 0x7E, data}
+	return r.spiDev.Tx(newData, nil)
+}
+
+func (r *Dev) devRead(address int) (byte, error) {
+	data := []byte{((byte(address) << 1) & 0x7E) | 0x80, 0}
+	out := make([]byte, len(data))
+	if err := r.spiDev.Tx(data, out); err != nil {
+		return 0, err
+	}
+	return out[1], nil
+}
+
+func (r *Dev) setBitmask(address, mask int) error {
+	current, err := r.devRead(address)
+	if err != nil {
+		return err
+	}
+	return r.devWrite(address, current|byte(mask))
+}
+
+func (r *Dev) clearBitmask(address, mask int) error {
+	current, err := r.devRead(address)
+	if err != nil {
+		return err
+	}
+	return r.devWrite(address, current&^byte(mask))
+}
+
+// Reset resets the RFID chip to initial state.
+func (r *Dev) Reset() error {
+	return r.devWrite(commands.CommandReg, commands.PCD_RESETPHASE)
+}
+
+// Halt soft-stops the chip - PowerDown bit set, command IDLE
+func (r *Dev) Halt() error {
+	return r.devWrite(commands.CommandReg, 16)
+}
+
+// SetAntenna configures the antenna state, on/off.
+func (r *Dev) SetAntenna(state bool) error {
+	if state {
+		current, err := r.devRead(commands.TxControlReg)
+		if err != nil {
+			return err
+		}
+		if current&0x03 != 0 {
+			return errors.New("Can not set the bitmask for antenna")
+		}
+		return r.setBitmask(commands.TxControlReg, 0x03)
+	}
+	return r.clearBitmask(commands.TxControlReg, 0x03)
+}
+
+// CardWrite the low-level interface to write some raw commands to the card.
+//
+// 	command - the command register
+// 	data - the data to write out to the card using the authenticated sector.
+func (r *Dev) CardWrite(command byte, data []byte) ([]byte, int, error) {
+	var backData []byte
+	backLength := -1
+	irqEn := byte(0x00)
+	irqWait := byte(0x00)
+
+	switch command {
+	case commands.PCD_AUTHENT:
+		irqEn = 0x12
+		irqWait = 0x10
+	case commands.PCD_TRANSCEIVE:
+		irqEn = 0x77
+		irqWait = 0x30
+	}
+
+	r.devWrite(commands.CommIEnReg, irqEn|0x80)
+	r.clearBitmask(commands.CommIrqReg, 0x80)
+	r.setBitmask(commands.FIFOLevelReg, 0x80)
+	r.devWrite(commands.CommandReg, commands.PCD_IDLE)
+
+	for _, v := range data {
+		r.devWrite(commands.FIFODataReg, v)
+	}
+
+	r.devWrite(commands.CommandReg, command)
+
+	if command == commands.PCD_TRANSCEIVE {
+		r.setBitmask(commands.BitFramingReg, 0x80)
+	}
+
+	i := 2000
```

Comment:
That is very adhoc and will have to be rewritten. IIUC, it should wait for the irqPin, no need to pool.

Reply:
This is the code I derived from the python library, so I'm not sure whether it's ad-hoc or not.

Reply:
I'd speculate it is. :) Let's revisit this later.